liqwid-plutarch-extra/src/Plutarch/Extra/StateThread.hs

```diff
@@ -54,8 +46,66 @@ pwithStateThread ::
   Term s (PMintingPolicy :--> PTxOutRef :--> PMintingPolicy)
 pwithStateThread = plam withStateThread
 
+{- | Adds a state thread to a minting policy
+ allowing more than one state thread token to be minted
+ Parameterized at the Haskell level.
+
+ @since 3.21.0
+-}
+withStateThreadMulti ::
+  forall (s :: S).
+  -- | Minting policy to wrap
+  Term s PMintingPolicy ->
+  -- | Initial spend
+  Term s PTxOutRef ->
+  Term s PMintingPolicy
+withStateThreadMulti = withStateThreadGeneric uniqueStateTokensMint
+
+{- | Adds a state thread to a minting policy
+ allowing more than one state thread token to be minted
+ Parameterized at the Plutarch level.
+
+ @since 3.21.0
+-}
+pwithStateThreadMulti ::
+  forall (s :: S).
+  Term s (PMintingPolicy :--> PTxOutRef :--> PMintingPolicy)
+pwithStateThreadMulti = plam withStateThread
+
 -- Helpers
 
+{- | Adds a state thread to a minting policy.
+ Parameterized at the Haskell level,
+ with parametrized check on minted assets.
+
+ @since 3.21.0
+-}
+withStateThreadGeneric ::
+  forall (s :: S).
+  ( forall (keys :: KeyGuarantees) (amounts :: AmountGuarantees).
+    Term s PCurrencySymbol ->
+    Term s (PValue keys amounts) ->
+    Term s PBool
+  ) ->
+  -- | Minting policy to wrap
+  Term s PMintingPolicy ->
+  -- | Initial spend
+  Term s PTxOutRef ->
+  Term s PMintingPolicy
+withStateThreadGeneric checkMint mp ref = plam $ \red ctx -> pletAll ctx $ \ctx' ->
+  pletFields @'["inputs", "mint"] (getField @"txInfo" ctx') $ \txInfo ->
+    pmatch (getField @"purpose" ctx') $ \case
+      PMinting thisPolicy ->
+        pif
+          (checkMint (pfield @"_0" # thisPolicy) . getField @"mint" $ txInfo)
+          ( pif
+              (pany # (hasUniqueInput # ref) # getField @"inputs" txInfo)
+              (mp # red # ctx)
```

Comment:
This should be `ctx'` I think.

Reply:
IMO, `ctx'` should be the non-pletted value, and `ctx` should be the pletted one